util/multirun.sh

```diff
@@ -35,27 +67,78 @@ fi
 
 cip="$1"
 shift
+args=("$@")
 
-for opts in "$@"; do
-    manifest=$(echo "$opts" | cut -d, -f1)
+for arg in "${args[@]}"; do
+    if ! [[ "$arg" =~ .*,.* ]]; then
+        echo >&2 "invalid argument: $arg"
+        usage
+        exit 1
+    fi
+done
+
+# If we only want to run against changed manifest files, then filter out those
+# files that were not changed since the last commit.
+if [[ "${CIP_GIT_REV:-}" ]]; then
+    if [[ ! -d "${CIP_GIT_DIR:-}" ]]; then
+        echo >&2 "CIP_GIT_DIR not set (must be set if CIP_GIT_REV is set)"
+        usage
+        exit 1
+    fi
+    pushd "${CIP_GIT_DIR}"
+    changed_files=()
+    while IFS= read -r f; do
+        changed_files+=( "$f" )
+    done < <( git diff-tree --no-commit-id --name-only -r "${CIP_GIT_REV}" )
+    popd
+
+    changed_manifests=$(collect_manifests "${cip}" "${changed_files[@]}")
+
+    args_filtered=()
+    for arg in "${args[@]}"; do
+        manifest=$(echo "$arg" | cut -d, -f1)
+        manifest_comparable=$(basename "${manifest}")
+        if echo "${changed_manifests[@]}" | grep "${manifest_comparable}"; then
+          args_filtered+=("${arg}")
+        fi
+    done
+
+    if (( "${#args_filtered[@]}" == 0 )); then
+        echo "No manifests were changed in ${CIP_GIT_REV}."
+        exit 0
+    fi
+
+    args=("${args_filtered[@]}")
+fi
+
+for arg in "${args[@]}"; do
+    manifest=$(echo "$arg" | cut -d, -f1)
     service_accounts=()
     service_account_keyfiles=()
     while IFS= read -r keyfile; do
-        service_account_keyfiles+=( "$keyfile" )
-    done < <( echo "$opts" | cut -d, -f2- | tr , '\n' )
-    for keyfile in "${service_account_keyfiles[@]}"; do
-        # Authenticate as the service account. This allows the promoter to later
-        # call gcloud with the flag `--account=...`. We can allow the service
-        # account creds file to be empty, for testing cip locally (for the case
-        # where the service account creds are already activated).
-        gcloud auth activate-service-account --key-file="${keyfile}"
-        service_accounts+=("$(gcloud config get-value account)")
-    done
+        if [[ -n "$keyfile" ]]; then
+            service_account_keyfiles+=( "$keyfile" )
+        fi
+    done < <( echo "$arg" | cut -d, -f2- | tr , '\n' )
```

Comment:
This seems baroque.  Is it simpler to do:

```
A="foo,bar,bat"
read manifest keys < <(echo "$arg" | sed 's/,/ /g')
keyfiles=($keys)
echo M: "$manifest", K: "${keyfiles[@]}"
```

Also doesn't require the trailing comma

Reply:
Oh nice --- I didn't realize `read` could take on multiple variable name args. Changing.

Reply:
I just opted for using 'readarray' which avoids the process substitution altogether.